lib/protocol/task.js

```diff
@@ -19,49 +19,47 @@ function taskProtocolFactory (Promise, assert, Constants, Errors, messenger, _,
     function TaskProtocol() {
     }
 
-    TaskProtocol.prototype.run = function (taskId, args) {
-        assert.uuid(taskId);
+    TaskProtocol.prototype.run = function (domain, data) {
+        assert.string(domain);
 
         return messenger.publish(
             Constants.Protocol.Exchanges.Task.Name,
-            'methods.run' + '.' + taskId,
-            new Result({ value: args })
+            domain + '.' + 'methods.run',
+            data || {}
         );
     };
 
-    TaskProtocol.prototype.subscribeRun = function(taskId, callback) {
-        assert.uuid(taskId);
+    TaskProtocol.prototype.subscribeRun = function(domain, callback) {
+        assert.string(domain);
         assert.func(callback);
 
         return messenger.subscribe(
             Constants.Protocol.Exchanges.Task.Name,
-            'methods.run' + '.' + taskId,
-            function(data) {
-                callback(data.value);
-            }
+            domain + '.' + 'methods.run',
+            callback
         );
     };
 
     TaskProtocol.prototype.cancel = function (taskId, errName, errMessage) {
         assert.uuid(taskId);
-        assert.string(errName);
-        assert.string(errMessage);
 
         return messenger.publish(
             Constants.Protocol.Exchanges.Task.Name,
-            'methods.cancel' + '.' + taskId,
-            { errName: errName, errMessage: errMessage }
+            'methods.cancel',
+            { taskId: taskId, errName: errName, errMessage: errMessage }
         );
     };
 
-    TaskProtocol.prototype.subscribeCancel = function(taskId, callback) {
-        assert.uuid(taskId);
+    TaskProtocol.prototype.subscribeCancel = function(callback) {
         assert.func(callback);
 
         return messenger.subscribe(
             Constants.Protocol.Exchanges.Task.Name,
-            'methods.cancel' + '.' + taskId,
+            'methods.cancel',
             function(data) {
+                if (!data.errName) {
+                    callback(data);
```

Comment:
Yep, that would be much better to exit the function here. It looks like we still won't evaluate the `if` statement below, so at least there's not an existing bug.